teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

```diff
@@ -171,12 +199,482 @@ In addition to grouping related resources, Tekton Results provides long term sto
+users to bundle `TaskRuns` and `Runs` to their parent `PipelineRuns`. It also provides
+long term storage of resources. Users can rely on [Tekton Results][results-api] to map
+`TaskRuns` and `Runs` to provide the mapping that was available in the full embedded
+statuses. 
+
+Note that Results API is still in Alpha, but progress is being made towards Beta - we
+estimate that the Results API will be in Beta by the time we remove the full embedded
+statuses.
+
+#### Tekton Dashboard
+
+[Tekton Dashboard][dashboard] shows the status the `TaskRuns` and `Runs` of a given
+`PipelineRun`, and this should continue to be supported. The Tekton Dashboard currently
+relies on the full embedded statuses, including when the scheduled cleanup of resources
+removed `TaskRuns` and `Runs` from the cluster. The Dashboard will need to be updated
+to use the minimal embedded statuses and rely on Tekton Results for long term storage
+(read more in [related issue][issue-82]).
+
+We expect the load on the Dashboard to reduce and its performance to improve, given
+that the `PipelineRuns` would not be reacting to the updates in `Steps`.
+
+#### Tekton Chains
+
+[Tekton Chain][chains] observes `TaskRuns` and signed them directly, they don't depend
+on the full embedded status in `PipelineRun` status, so they should be unaffected. When
+Tekton Chains starts to sign `PipelineRuns`, as proposed in [TEP-0084][tep-0084], we are
+exploring creating a single attestation record upon completion of a `PipelineRun` that
+includes everything - `TaskRuns`, `PipelineRun`, and `event-payload` - instead of a record
+for each of them. We will ensure that the proposal in TEP-0084 aligns with the changes to
+`PipelineRuns` proposed in this TEP.
+
+## Design Evaluation
+
+1. **API conventions**: This design complies with the [Kubernetes API conventions][subobjects]
+   by using sub-objects instead of maps for fields, and using string aliases instead of booleans
+   for behavior flags.
+2. **Simplicity**: This design simplifies the `PipelineRun` status by providing the minimum
+   information and updates needed from `TaskRuns` and `Runs`.
+3. **Reusability**: This design encourages reuse of existing components, such as Owner References
+   and Tekton Results, by removing the duplication caused by full embedded statuses.
+4. **Flexibility**: This design improves the extensibility of Tekton Pipelines to support upcoming
+   features that create multiple `TaskRuns`, `Runs` or `PipelineRuns` from a single `PipelineTask`.
+   The behavior flag is also flexible to support more configurations if needed.
+5. **Conformance**: This design impact the conformance surface through changes to the `PipelineRun`
+   interface. The changes are backwards incompatible but will be introduced in a backwards compatible
+   manner first with migration instructions and deprecation warnings.
+
+## Alternatives
+
+### Add Minimal Embedded Status - Collapse Fields
```

Comment:
Opening another comment thread here since GitHub PRs don't allow reopening outdated comment threads. 😢  (original thread - https://github.com/tektoncd/community/pull/616#discussion_r796785042)

We had some additional discussion about this alternative that we were somewhat split on. Would love additional input here!

The rough idea is to reference the sub-resources with TypeMeta+ObjectMeta to be in line with [multiple resource reference recommendations](https://github.com/tektoncd/pipeline/issues/4543) and give us a way to handle arbitrary types (e.g. TaskRun, Run, PipelineRun(?)), but also a way for us to distinguish versions of the same resource (e.g. v1 vs v1beta1 Tasks) (related discussion in https://github.com/tektoncd/pipeline/issues/4543 for how we want to handle mixed apiVersion definitions). e.g.

```go
type ResourceStatus struct {
	runtime.TypeMeta
	metav1.ObjectMeta
	duckv1beta1.Status
}
```

Even though it's slightly more verbose, it gives us more flexibility in the types we can support now and in the future.

The main downside is by using the same message type, we're making it more difficult to add type-specific fields in the future (unless we use some sort of duck typing). However, since minimal statuses are intended to be ref-like objects anyway, it's unclear if we will actually need this.

Let us know what you think!

Reply:
I like the idea of including a list of resources compared to a map of `taskRun` and `runs` https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#monitoring-execution-status. This demands a little more filtering to get to the individual status compared to what we have today. But give the extensibility it provides and we are at the stage of revamping the status anyways, its worth exploring this option.